### ISSUE TITLE
Drag between two sortables

### ISSUE THREAD

**Issue by LukasUrban** [gaearon/react-dnd#112]
I'm having problem moving items between two different zones. It is possible to move item from one to another, but then new drag has to be initiated to sort in the new location.

Btw is it somehow possible to drag item to empty zone?

Live example - http://fluid.cz/projects/react-dnd/

```
function _moveWidget(e){
	var overComponent = e.overComponent.props
	var draggedComponent = e.draggedComponent.props

	var draggedComponentObj = _.find(_page.zones[draggedComponent.zoneKey].widgets, {'id':draggedComponent.widgetData.id}) 
	var overComponentObj = _.find(_page.zones[overComponent.zoneKey].widgets, {'id':overComponent.widgetData.id}) 

	draggedIndex = _page.zones[draggedComponent.zoneKey].widgets.indexOf(draggedComponentObj);
	overIndex = _page.zones[overComponent.zoneKey].widgets.indexOf(overComponentObj);
	
	_page.zones[draggedComponent.zoneKey].widgets.splice(draggedIndex,1)
	_page.zones[overComponent.zoneKey].widgets.splice(overIndex, 0, draggedComponentObj)
}
```

**Comment by gaearon** [comment]
To implement dragging to an empty zone, make that zone a drop target.

=====

As for troubles with dragging from one list to another: I see a lot of JS errors in console when I do that.

Specifically, these ones:

![screen shot 2015-03-10 at 19 52 59](https://cloud.githubusercontent.com/assets/810438/6580042/13b4a260-c75f-11e4-83a8-06141369b20e.png)


The problem seems to be the following: you look for the widget info **in the original zone**:

```js
var draggedComponentObj = _.find(_page.zones[draggedComponent.zoneKey].widgets, {'id':draggedComponent.widgetData.id}) 
```

However, if the zone has changed, you can no longer find it there! Hence `draggedComponentObj` being `null` and an exception.

What you should do instead is to **store items separately by ID**, not inside `zones`. You can have something like flat `_page.widgets` (I wouldn't use term “component”, it conflicts with React's meaning for component). You wouldn't even need `find` this way:

```js
var draggedComponentObj = _page.widgets[draggedComponent.widgetData.id]; 
```

This way it will be findable even if the zone has changed.

=====

Finally, you'll face troubles with `getDragState(type).isDragging` returning `false` once component has been “moved” to a different zone (since technically it's a different component). This is a known issue related to current React DnD design and will be resolved in big 0.10 release in the following month.

**Comment by LukasUrban** [comment]
Thanks for lightning fast and very complex answer. Now I pretty much get everything except proposed structure of items. 

You probably do not mean object with ids as keys, do you? 
```
{uid:{widgetData}, uid2:{otherWidgetData}}
```

**Comment by gaearon** [comment]
Why not? Pretty much what I meant.

**Comment by LukasUrban** [comment]
Array of objects seems much more natural to me, but I'm going to investigate that.

**Comment by gaearon** [comment]
It's up to you but a map seems more natural here because there is no defined order (it's just a mapping from ID to instance). In JS you don't have native `Map`s yet so an object is as close as you can get.

**Comment by gaearon** [comment]
Closing, as it's more of a question than an issue.

As I said before, we're planning to solve “wrong `isDragging`” problem for “resurfacing” components (such as when widget mounts in another list). You can track progress in #111.

**Comment by ntdb** [comment]
Hey @gaearon, has the `isDragging` problem been addressed? I just upgraded to 1.0.0, rewired everything for the new API, and `isDragging` still seems to be wrong after a component is dragged to a new list.

**Comment by gaearon** [comment]
@ntdb Thanks for using it! Yeah, `isDragging` is the way to go.
